src/telemetry/Manager.h

```diff
@@ -50,6 +58,101 @@ class Manager
 	 */
 	virtual void InitPostScript();
 
+	/**
+	 * Instances returned by CollectMetrics().
+	 */
+	enum MetricType
+		{
+		DOUBLE_COUNTER,
+		INT_COUNTER,
+		DOUBLE_GAUGE,
+		INT_GAUGE,
+		DOUBLE_HISTOGRAM,
+		INT_HISTOGRAM,
+		};
+
+	struct CollectedMetric
+		{
+
+		CollectedMetric(Manager::MetricType metric_type,
+		                const broker::telemetry::metric_family_hdl* family,
+		                std::vector<std::string_view> label_values, std::optional<double> value,
+		                std::optional<int64_t> count_value)
+
+			: metric_type(metric_type), family(family), label_values(std::move(label_values)),
+			  value(value), count_value(count_value)
+			{
+			}
+
+		enum MetricType metric_type;
+		const broker::telemetry::metric_family_hdl* family;
+		std::vector<std::string_view> label_values;
+		std::optional<double> value;
+		std::optional<int64_t> count_value;
+
+		/**
+		 * Create a Metric record.
+		 */
+		zeek::RecordValPtr GetMetricRecord() const;
+		};
+
+	struct CollectedHistogramMetric
+		{
+		CollectedHistogramMetric(MetricType metric_type,
+		                         const broker::telemetry::metric_family_hdl* family,
+		                         std::vector<std::string_view> label_values, int64_t count,
+		                         std::vector<double> values, double sum,
+		                         std::optional<std::vector<int64_t>> count_values,
+		                         std::optional<int64_t> count_sum, std::vector<double> bounds,
+		                         std::optional<std::vector<int64_t>> count_bounds)
+
+			: metric_type(metric_type), family(family), label_values(std::move(label_values)),
+			  count(count), values(std::move(values)), sum(sum),
+			  count_values(std::move(count_values)), count_sum(count_sum),
+			  bounds(std::move(bounds)), count_bounds(std::move(count_bounds))
+			{
+			}
+
+		enum MetricType metric_type;
+		const broker::telemetry::metric_family_hdl* family;
+		std::vector<std::string_view> label_values;
+
+		int64_t count; // Observations are counted full, use int64_t
+		std::vector<double> values;
+		double sum;
+		std::optional<std::vector<int64_t>> count_values;
+		std::optional<int64_t> count_sum;
+
+		// The following two should be the same for every metric
+		// within a family, but the broker::telemetry API doesn't
+		// allow us to query bounds/buckets using the family, so
+		// for now get those from the instances.
+		std::vector<double> bounds;
+		std::optional<std::vector<int64_t>> count_bounds;
```

Comment:
When using the `variant`-based approach from above, the members would boil down to:

```
		const broker::telemetry::metric_family_hdl* family;
		std::vector<std::string_view> label_values;
		std::variant<IntHistogram, DblHistogram> data;
```

Storing a `MetricType` is unnecessary (and even counterproductive) since we can look at the type of `data` to figure out which `MetricType` this correlates to.